### PR TITLE
ruby 3.1.2

### PR DESCRIPTION
**Pull request by betmenslido**
Lets get latest stable ruby version packaged :)

I made a few changes to get 3.1.2 build working, and also did some cleanup.

Few patches were commented out for a long time so I removed them.

**0001**-mingw-w64-time-functions
disabled 3y ago (ruby 2.6.3): https://github.com/msys2/MINGW-packages/commit/18b746389f4c84785fdde13f37cfac5e7e4bb4cd

**004**-rbinstall-destdir.patch
disabled 5y ago (ruby 2.4.0)
https://github.com/msys2/MINGW-packages/commit/53a544548065c88301db50e21fff3bb7051bae24

**0005**-rbinstall-rake-bash.patch
Fixed upstream. I double checked and this change is in ruby-3.1.2 sources.
https://github.com/ruby/ruby/commit/8544f51ef70b6b34aa6c81f5b52a4f05da464609

**0006**-autoconf-set-msvcrt-ver-for-ucrt.patch
AFAIU ucrt build should be working out of box (and worked for me) because ruby upstream explicitly supports it and there are CI workflow for it.
* https://github.com/ruby/ruby/pull/4599
* CI example for 3.1 branch commit: https://github.com/ruby/ruby/actions/runs/2663213558
* RubyInstaller 3.1 version also uses ucrt, and there are no patches related to ucrt functioning. https://github.com/oneclick/rubyinstaller2-packages/tree/master/mingw-w64-ruby31

**0008**-ucrt-32-win11-insns.patch
There are fix upstream which fixes what appears to be typo (and causes conflict).
https://github.com/ruby/ruby/commit/7c31ecd3accd75d164b4f87b0651b47f31da286d
I fixed patch to restore previous behaviour, but I am not 100% sure about this change.
AFAIU this is for clang32 build on win11. It would be awesome if someone could help test this particular build, because at the moment I dont have win11 VM ready.

Also, in PKGBUILD file itself 2 fixes were made:
1) Only gem script is copied to final package because other scripts are generated elsewhere now
2) rake.bat fixup is no longer needed - i checked multiple builds and generated rake.bat/rake combo is fine - bat file references ruby from current dir.
f:\rubies\3.1.2-clang32\bin>rake.bat --version
rake, version 13.0.6

### CODE REVIEW

**Comment by betmenslido** [comment]
So I did few tests on win11, and indeed without patch 008 clang32 executable faults with "unexpected ucrtbase.dll". Patched version worked fine. Being curious, I also used subset of tests shipped with source archive ("ruby-3.1.2.tar.gz\ruby-3.1.2\test\") to test clang32 executable on win11 and did not notice any mis-behaviour (compared to same tests being executed with RubyInstaller 3.1.2 distribution).

File: mingw-w64-ruby/0008-ucrt-32-win11-insns.patch [hunk 1]
```diff
@@ -15,7 +15,7 @@
 +        for (pend += 10; pend < p + 500; pend++) {
              // find end of function
 -            if (memcmp(pend, FUNCTION_BEFORE_RET_MARK, sizeof(FUNCTION_BEFORE_RET_MARK) - 1) == 0 &&
--                *(pend + (sizeof(FUNCTION_BEFORE_RET_MARK) - 1) + FUNCTION_SKIP_BYTES) & FUNCTION_RET == FUNCTION_RET) {
+-                (*(pend + (sizeof(FUNCTION_BEFORE_RET_MARK) - 1) + FUNCTION_SKIP_BYTES) & FUNCTION_RET) == FUNCTION_RET) {
```
Comment (betmenslido) [inline mingw-w64-ruby/0008-ucrt-32-win11-insns.patch:18]: Apart from modified line numbers, only parentheses were added to make this patch compatible with new win32.c version.

File: mingw-w64-ruby/PKGBUILD [hunk 2]
```diff
@@ -214,15 +195,10 @@ check() {
...
-  for script in {erb,gem,irb,rdoc,ri}; do
+  for script in {gem,}; do
     install ${srcdir}/ruby-${pkgver}/bin/${script} \
       ${pkgdir}${MINGW_PREFIX}/bin/
   done
-  local PREFIX_WIN
-# fixup rake.bat so that it can be invoked from "cmd"
-  PREFIX_WIN=$(cygpath -wm ${pkgdir}${MINGW_PREFIX})
-  sed -e "s|\\\|/|g" -e "s|${MINGW_PREFIX}/bin/ruby.exe|ruby.exe|g" \
-      -e "s|${PREFIX_WIN}/bin/rake|rake|g" -i "${pkgdir}${MINGW_PREFIX}/bin/rake.bat"
```
Comment (betmenslido) [inline mingw-w64-ruby/PKGBUILD:225]: Final packaged rake.bat+rake combo was correct in all builds, no need to patch anything.

**Comment by betmenslido** [comment]
So I did few tests on win11, and indeed without patch 008 clang32 executable faults with "unexpected ucrtbase.dll". Patched version worked fine. Being curious, I also used subset of tests shipped with source archive ("ruby-3.1.2.tar.gz\ruby-3.1.2\test") to test clang32 executable on win11 and did not notice any mis-behaviour (compared to same tests being executed with RubyInstaller 3.1.2 distribution).

**Comment by mmuetzel** [comment]
Nice work!

The MINGW builds seem to depend on libgmp. The CLANG builds don't. Would it make sense to add `"${MINGW_PACKAGE_PREFIX}-gmp"` to the dependencies to have the CLANG builds also use features from that library?

The package grokker says that `mingw-w64-ngraph-gtk` needs a pkgrel bump with this update.

**Comment by lazka** [comment]
Thanks for the detailed description of the changes.

**Pull request merged**